### PR TITLE
Don't allow nested argument constraints

### LINKED ISSUE #1177: Argument constraint That (or Ignored, or _), when nested deeper in A.CallTo, misreports what's being matched
Found while investigating  [FakeItEasy - How to verify nested arguments value C#](https://stackoverflow.com/questions/45411178/fakeiteasy-how-to-verify-nested-arguments-value-c-sharp). 

Consider this test:

```c#
public class NestedThat
{
    public class Dto
    {
        public string Id;
    }

    public interface IService
    {
        void DoIt(Dto dto);
    }

    [Test]
    public void Test()
    {
        var fake = A.Fake<IService>();
        fake.DoIt(new Dto { Id = "some id" });

        A.CallTo(() => fake.DoIt(new Dto { Id = A<string>._ })).MustHaveHappened();
    }
}
```

The author would expect a passing test, but it fails, which isn't horrible, since the use of `A<string>._` inside the `Dto constructor` is illegal. However, the failure does not lead the reader to the fact that the use of `._` (and I'm sure any variant of `That`) is bad. Here's the output:

```
Test 'FakeItEasyPlay.NestedThat.Test' failed: FakeItEasy.ExpectationException : 

  Assertion failed for the following call:
    FakeItEasyPlay.NestedThat+IService.DoIt(dto: <Ignored>)
  Expected to find it at least once but found it #0 times among the calls:
    1: FakeItEasyPlay.NestedThat+IService.DoIt(dto: FakeItEasyPlay.NestedThat+Dto)

	at FakeItEasy.Core.FakeAsserter.AssertWasCalled(Func`2 callPredicate, String callDescription, Func`2 repeatPredicate, String repeatDescription)
	at FakeItEasy.Configuration.RuleBuilder.MustHaveHappened(Repeated repeatConstraint)
	at FakeItEasy.AssertConfigurationExtensions.MustHaveHappened(IAssertConfiguration configuration)
	NestedThat.cs(19,0): at FakeItEasyPlay.NestedThat.Test()
```

In addition to not highlighting the illegal `._`, we are misreporting the expected argument to `DoIt` as being equivalent to `A<Dto>._`.

It may be difficult (impossible) to throw correct error in all cases, but in this one, where we're capturing  a `That` constraint to match against a parameter of a different type, I'd hope we'd be able to do _something_.

### PR DESCRIPTION
**Pull request by thomaslevesque**
Fixes #1177

### CODE REVIEW

File: tests/FakeItEasy.Specs/CallMatchingSpecs.cs [hunk 1]
```diff
@@ -498,6 +497,76 @@ public static void UnusedArgumentMatcherDescriptionNotUsed(IHaveNoGenericParamet
...
+        }
+
+        [Scenario]
+        public static void PassingNestedUnderscoreConstraintToAMethod(
+            IHaveOneGenericParameter fake, Exception exception)
+        {
+            "Given a fake"
+                .x(() => fake = A.Fake<IHaveOneGenericParameter>());
+
+            "When I try to configure a method of the fake with a Ignored constraint nested in an argument"
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/CallMatchingSpecs.cs]: Ah, yes... I copied & pasted, and forgot to change the text

File: tests/FakeItEasy.Specs/CallMatchingSpecs.cs [hunk 2]
```diff
@@ -498,6 +497,76 @@ public static void UnusedArgumentMatcherDescriptionNotUsed(IHaveNoGenericParamet
...
+        }
+
+        [Scenario]
+        public static void PassingNestedThatMatchesConstraintToAMethod(
+            IHaveOneGenericParameter fake, Exception exception)
+        {
+            "Given a fake"
+                .x(() => fake = A.Fake<IHaveOneGenericParameter>());
+
+            "When I try to configure a method of the fake with a Ignored constraint nested in an argument"
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/CallMatchingSpecs.cs]: Indeed!

**Comment by thomaslevesque** [comment]
I pushed a fixup commit.

> I'm not sure I follow the very fine details for the expression walking

The visitor visits the expression by "drilling down" its nodes. If I find the `A.That.Matches(...)` node as an argument constraint, I will also find the `A.That` node later. Since I need to walk the call chain to find if something is called on `A<T>`, I keep track of the nodes I already checked, so that I don't report them again as argument constraints when these nodes are visited later.

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs [hunk 3]
```diff
@@ -106,5 +116,95 @@ private IArgumentConstraint CreateParamArrayConstraint(NewArrayExpression expres
 
             return new AggregateArgumentConstraint(result);
         }
+
+        private class ArgumentConstraintExpressionVisitor : ExpressionVisitor
+        {
+            private readonly List<Expression> constraintExpressions;
+            private readonly HashSet<Expression> visitedExpressions;
```
Comment (blairconrad) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: You did. Thanks! Postorder traversal. I didn't realize.

**Comment by blairconrad** [comment]
> I pushed a fixup commit.

Yeah! It was nice.

> > I'm not sure I follow the very fine details for the expression walking

> The visitor visits the expression by "drilling down" its nodes. If I find the `A.That.Matches(...)` node as an argument constraint, I will also find the `A.That` node later. Since I need to walk the call chain to find if something is called on `A<T>`, I keep track of the nodes I already checked, so that I don't report them again as argument constraints when these nodes are visited later.

Yeah. It's more the bouncing between invoking on members and on methods and just on anything. I believe in it.

**Comment by blairconrad** [comment]
Oh, yeah. 

1. analyzer - do you want a different issue?
2. docs

**Comment by thomaslevesque** [comment]
Looks like I missed some of your comments... I'll look into it tomorrow.

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs [hunk 4]
```diff
@@ -80,8 +78,20 @@ private static object InvokeExpression(Expression expression)
             return Expression.Lambda(expression).Compile().DynamicInvoke();
         }
 
+        private static void CheckArgumentExpressionIsValid(Expression expression)
+        {
+            var visitor = new ArgumentConstraintExpressionVisitor();
+            visitor.Visit(expression);
+            if (visitor.ConstraintExpressions.Any(e => e != expression))
+            {
+                throw new InvalidOperationException(ExceptionMessages.ArgumentConstraintCannotBeNestedInArgument);
```
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: That was my initial intention, but then I realized that getting a sensible representation of the expression was far from trivial; we don't have access to the original source code, as in the analyzer. And anyway, should we show the expression in C#, VB, something else?

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs [hunk 5]
```diff
@@ -106,5 +116,95 @@ private IArgumentConstraint CreateParamArrayConstraint(NewArrayExpression expres
 
             return new AggregateArgumentConstraint(result);
         }
+
+        private class ArgumentConstraintExpressionVisitor : ExpressionVisitor
+        {
+            private readonly List<Expression> constraintExpressions;
```
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: Yes, that's an option too. We could even throw as soon as we find an invalid argument constraint. I can change it if you prefer.

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs (hunk 3 above)
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: > Postorder traversal

If you say so ;)

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs (hunk 4 above)
Comment (blairconrad) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: I'm not sure how easy it is to get a sensible representation of the expression (but I'd think that extracting out the `That`, `_`, or `Ignored` would be sufficient). As for language, if just extracting the member name isn't enough (or isn't feasible), I don't suppose getting the "CLR formatting" is an option with an expression, is it? Like we did in #1047. Perhaps one would have to compile the method to do so?

As a cheap alternative, we could augment the static exception message we have above. Maybe something like 

>An argument constraint, such as That, Ignored, or _, cannot be nested in an argument.

But if we have a `That` inside a `That`, maybe it doesn't help much.

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs (hunk 5 above)
Comment (blairconrad) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: I just thought it would leave us with slightly simpler code and would be (probably imperceptibly) faster in the (very rare) case where a user makes this error. Your choice.

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs (hunk 3 above)
Comment (blairconrad) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: I did, but I'm not so sure now. It may be someotherorder traversal, but you double visit because of the shenanigans that I didn't entirely understand - you're forcing a descent down the tree in some cases.

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs (hunk 4 above)
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: `Expression.ToString()` returns things like ``A`1.That.Matches(x => True)``, which isn't great, but is consistent with the CLR formatting used elsewhere... But when we do the same from VB.NET, we get something like this: ``() => Convert(A`1.That).Matches(a0 => Invoke(() => True))``, which is much uglier.

The cheap alternative sounds reasonable to me.

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs (hunk 3 above)
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: > you're forcing a descent down the tree in some cases.

Yes. But now that I think about it, there might be an easier way. Let me try something.

Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: OK, now I avoid the double visit. Better?

**Comment by thomaslevesque** [comment]
> analyzer - do you want a different issue?

Yes, I think we can treat it separately. I'll open an issue.

>     docs

I just pushed a new commit to mention the restriction on nested argument constraints.

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque. I like the docs. The examples really help illustrate a concept that's hard to describe.

Sorry for the delay. I've been busy, but also struggling to understand the code. I believe in it (feel free to squash at any time), but worry that it's hard to comprehend, and to modify.
I was debugging through to try to understand how it worked, and while I was doing that, I tried a different approach. Would you mind taking a look at it? To me, it's easier to understand, but maybe it's just me.
https://github.com/blairconrad/FakeItEasy/tree/fix-nested-arg-constraint

**Comment by thomaslevesque** [comment]
Well, I kind of liked my approach, but it's true that yours is easier to understand, which is probably more important in the long run. Do you want to push it to my branch, or should I fixup my commits?

**Comment by blairconrad** [comment]
> I kind of liked my approach

I know. I feel bad saying "try it this way instead", but I did find the logic hard to follow.
If you want to keep your code, perhaps you could add a comment explaining exactly 
why the `candidateExpression` ends up being the value we need. I _think_ it's because `base.VisitMethodCall` traverses the method call's arguments, but we then exit `NestedArgumentConstraintExpressionVisitor. VisitMethodCall` before `ExpressionVisitor` visits the `A<T>.That` (in `A<T>.That.Matches). But even writing this, I'm not sure.

I'll push an update to my commit (I saw a typo). If you want to take my code, I don't care whether you cherry-pick or just take the content and improve it or whatever. Commit credit isn't important. :wink:

Oh. While reading this morning, I saw some comments I'd make on your original code… I'm rushing out the door, but I'll try to get a few in, in case you want to keep yours.

**Comment by thomaslevesque** [comment]
> Oh. While reading this morning, I saw some comments I'd make on your original code… I'm rushing out the door, but I'll try to get a few in, in case you want to keep yours.

Don't bother, I'm convinced by your solution. Thinking about it again, I realize I no longer understand mine, which is a pretty clear sign that it'll be hard to maintain.

**Comment by blairconrad** [comment]
@thomaslevesque, thanks for the updates. I'm happy to merge after a rebase!

**Comment by thomaslevesque** [comment]
> I'm happy to merge after a rebase!

Done!

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque!

**Comment by thomaslevesque** [comment]
Thanks for the review and suggestions 👍
